cfme/tests/infrastructure/test_discover_multiple_providers.py

```diff
@@ -0,0 +1,116 @@
+import pytest
+from itertools import combinations
+
+from utils import testgen
+from utils.providers import get_crud, clear_infra_providers
+from cfme.infrastructure.provider import (discover, RHEVMProvider, SCVMMProvider, VMwareProvider)
+
+
+def generate_signature(combination):
+    return '-'.join([p.type for p in combination])
+
+
+def find_neighbour_provider_combinations(providers, upto):
+    neighbours = []
+    combinations_seen = set()
+
+    for count in range(1, upto + 1):
+        for combination in combinations(providers, count):
+            ip_prefix = '.'.join(combination[0].ip_address.split('.')[:3])
+            if not all(provider.ip_address.startswith(ip_prefix) for provider in combination[1:]):
+                continue
+            # They are in one /24, which is good
+            # Filter out duplicite provider types
+            filtered_combination = []
+            types_seen = set()
+            for provider in combination:
+                if provider.type in types_seen:
+                    continue
+                types_seen.add(provider.type)
+                filtered_combination.append(provider)
+            # We have ensured the provider types are unique - appearing only once
+            combination = sorted(filtered_combination, key=lambda provider: provider.type)
+            # Check if we already have such combination in place, if yes, skip this round
+            combination_tuple = tuple(provider.type for provider in combination)
+            if combination_tuple in combinations_seen:
+                continue
+            combinations_seen.add(combination_tuple)
+            # Finally, add this parametrization in place
+            neighbours.append(combination)
+
+    return neighbours
+
+
+def minmax_ip(providers):
+    ips = sorted([tuple(map(int, provider.ip_address.split('.'))) for provider in providers])
+    return '.'.join(map(str, ips[0])), str(ips[-1][-1])   # Last number of last IP
+
+
+def pytest_generate_tests(metafunc):
+    types = ['virtualcenter', 'rhevm', 'scvmm']
+    argnames, argvalues, idlist = testgen.provider_by_type(
+        metafunc, types)
+
+    argnames = ['providers_for_discover', 'start_ip', 'max_range']
+    new_id_list = []
+
+    providers_complete = []
+    providers_final = []
+
+    for x in idlist:
+        providers_complete.append(get_crud(x))
+
+    provider_combinations = sorted(
+        find_neighbour_provider_combinations(providers_complete, len(types)), key=len)
+    signatures_seen = set()
+
+    for prov_comb in provider_combinations:
+        sig = generate_signature(prov_comb)
+        if sig in signatures_seen:
+            continue
+        signatures_seen.add(sig)
+        start_ip, max_range = minmax_ip(prov_comb)
+        providers_final.append([prov_comb, start_ip, max_range])
+        new_id_list.append(sig)
+
+    testgen.parametrize(metafunc, argnames, providers_final, ids=new_id_list, scope="module")
+
+
+@pytest.yield_fixture(scope='module')
```

Comment:
I think this should be `"function"` scoped, anyway I think this is good to be merged.